src/realm/views.hpp

```diff
@@ -30,46 +30,104 @@ const int64_t detached_ref = -1;
 
 class RowIndexes;
 
-// SortDescriptor encapsulates a reference to a set of columns (possibly over links), which is
-// used to indicate the criteria columns for sort and distinct. Although the input is column
-// indices, it does not rely on those indices remaining stable as long as the columns continue to exist.
-class SortDescriptor {
+// Forward declaration needed for deleted CommonDescriptor constructor
+class SortDescriptor;
+
+// CommonDescriptor encapsulates a reference to a set of columns (possibly over
+// links), which is used to indicate the criteria columns for sort and distinct.
+// Although the input is column indices, it does not rely on those indices
+// remaining stable as long as the columns continue to exist.
+class CommonDescriptor {
 public:
-    SortDescriptor() = default;
-    SortDescriptor(SortDescriptor const&) = default;
-    SortDescriptor(SortDescriptor&&) = default;
-    SortDescriptor& operator=(SortDescriptor const&) = default;
-    SortDescriptor& operator=(SortDescriptor&&) = default;
+    CommonDescriptor() = default;
+    // Enforce type saftey to prevent automatic conversion of derived class
+    // SortDescriptor into CommonDescriptor at compile time.
+    CommonDescriptor(const SortDescriptor&) = delete;
+    virtual ~CommonDescriptor() = default;
 
-    // Create a sort descriptor for the given columns on the given table.
+    // Create a descriptor for the given columns on the given table.
     // Each vector in `column_indices` represents a chain of columns, where
     // all but the last are Link columns (n.b.: LinkList and Backlink are not
     // supported), and the final is any column type that can be sorted on.
     // `column_indices` must be non-empty, and each vector within it must also
-    // be non-empty. `ascending` must either be empty or have one entry for each
-    // column index chain.
-    SortDescriptor(Table const& table, std::vector<std::vector<size_t>> column_indices,
-                   std::vector<bool> ascending = {});
+    // be non-empty.
+    CommonDescriptor(Table const& table, std::vector<std::vector<size_t>> column_indices);
+    virtual CommonDescriptor* clone() const;
```

Comment:
It'd be safer to have `clone` return a `std::unique_ptr<CommonDescriptor>` to prevent a caller from forgetting to adopt the object into a `unique_ptr`.